syntax = "proto3";

import "proto/modules.proto";
import "proto/transaction.proto";

package slog.internal;

message Replica {
    // List of all server addresses in the system.
    // This list must have the size equal to number of partitions
    // If protocol is "tcp", these are IP addresses.
    // If protocol is "icp", these are filesystem paths.
    repeated string addresses = 1;
    // AWS public addresses for the servers. This field is only used by the admin tool.
    // If not specified, the addresses field is used instead.
    repeated string public_addresses = 2;
    // List of machines used for benchmarking. This field is only used by the admin tool
    repeated string client_addresses = 3;
    // Comma-separated string of replica id (e.g "1,3,2") ordered from closest replica to
    // furthest replica, not including this replica. This field is only used by the admin tool.
    string distance_ranking = 4;
}

message ReplicationDelayExperiment {
    // Percent of batches that are delayed
    uint32 delay_pct = 1;
    // Average duration of delay in milliseconds
    uint32 delay_amount_ms = 2;
}

/**
 * With hash partitioning, each key is interpreted as a byte string.
 * The keys are distributed to the partitions based on their
 * hash values.
 * Initial data can be loaded from an external file.
 */
message HashPartitioning {
    // Number of prefix bytes of a key to use for partitioning
    uint32 partition_key_num_bytes = 1;
}

/**
 * With simple partitioning, each key is assumed to be a number.
 * The keys are distributed to the partitions based on key values
 * modulo the number of partitions. 
 * Initial data cannot be loaded from an external file but can be
 * generated at startup. The following is an example of home and partition
 * assignment of the generated data with 3 partitions and 4 regions.
 *
 *        home | 0  1  2  3  0  1  2  3  0  ...
 * ------------|-------------------------------
 * partition 0 | 0  3  6  9  12 15 18 21 24 ...
 * partition 1 | 1  4  7  10 13 16 19 22 25 ...
 * partition 2 | 2  5  8  11 14 17 20 23 26 ...
 * ------------|-------------------------------
 *             |            keys            
 */
message SimplePartitioning {
    // Number of records to generate
    uint64 num_records = 1;
    // Size of a generated record in bytes
    uint32 record_size_bytes = 2;
}

/**
 * Similar to simple partitioning but instead of using the whole key, only the warehouse id
 * portion of the key is used for partitioning.
 */
message TPCCPartitioning {
    // Number of warehouses
    int32 warehouses = 1;
}

message CpuPinning {
    ModuleId module = 1;
    uint32 cpu = 2;
}

enum ExecutionType {
    KEY_VALUE = 0;
    NOOP = 1;
    TPC_C = 2;
}

/**
 * The schema of a configuration file.
 */
message Configuration {
    // Protocol for the zmq sockets in the broker. Use "tcp" for
    // normal running and "icp" for unit and integration tests
    string protocol = 1;
    // Replica groups. Each group has a list of machine addresses
    // with the size equal to number of partitions
    repeated Replica replicas = 2;
    // Port for the zmq sockets in the broker. Each value will create a new broker thread
    repeated uint32 broker_ports = 3;
    // Port for the server
    uint32 server_port = 4;
    // Port for the forwarder
    uint32 forwarder_port = 5;
    // Port for the sequencer
    uint32 sequencer_port = 6;
    // Number of partitions per replica
    uint32 num_partitions = 7;
    // Partitioning scheme
    oneof partitioning {
        HashPartitioning hash_partitioning = 8;
        SimplePartitioning simple_partitioning = 9;
        TPCCPartitioning tpcc_partitioning = 10;
    }
    // Number of worker threads for processing the transactions
    uint32 num_workers = 11;
    // How long the forwarder waits for batching
    uint64 forwarder_batch_duration = 12;
    // How long the sequencer waits for batching
    uint64 sequencer_batch_duration = 13;
    // Number of replicas that need to be synchronously replicated to
    uint32 replication_factor = 14;
    // Order of replicas for synchronous replication. This is a list of strings each of which corresponds
    // to a replica. Each string is an ordered comma-separated list of replica ids that the current replica
    // will synchronously replicate to based on replication order. The current replica will be ignored if exists
    // in the list. For example, ['1,2', '2,1,0', ''] means that, with a replication factor of 3, replica 0 will
    // synchronously replicate to 1 and 2, replica 1 will replicate to 2 and 0, and replica 2 will not synchronously
    // replicate to anywhere.
    repeated string replication_order = 15;
    // Replication of txn batches will be delayed to simulate uneven network latency
    ReplicationDelayExperiment replication_delay = 16;
    // Enable recording for the specified events
    repeated TransactionEvent enabled_events = 17;
    // For multi-home txn, send lock-only txns directly to the regions, skipping the global orderer
    bool bypass_mh_orderer = 18;
    // Pin each module to a cpu
    repeated CpuPinning cpu_pinnings = 19;
    // Return dummy txns back to the client instead of full results
    bool return_dummy_txn = 20;
    // Number of retries after calling zmq::recv something
    int32 recv_retries = 21;
    // Way of executing code of a txn
    ExecutionType execution_type = 22;
    // For each MH transaction, the involved regions will wait in such a way that their respective lock-only 
    // txns are put into the batches at approximately the same time
    bool synchronized_batching = 23;
    // Percentage of metrics to sample
    uint32 sample_rate = 24;
}